### ISSUE TITLE
class

### ISSUE THREAD

**Issue by pk1762012** [pk1762012/llm_pratik#4]
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession, functions as F, Window
from pyspark.ml.feature import VectorAssembler
from xgboost.spark import SparkXGBClassifier, SparkXGBClassifierModel
import matplotlib.pyplot as plt
import time

class BivariateAnalysis:
    def __init__(self, app_name):
        self.app_name = app_name
        self.spark = self._initialize_spark()

    def _initialize_spark(self):
        conf = SparkConf().setAppName(self.app_name).setMaster("yarn")  # Add other Spark configurations here
        sc = SparkContext(conf=conf).setLogLevel("ERROR")
        spark = SparkSession.builder.config(conf=conf).enableHiveSupport().getOrCreate()
        return spark

    def display_dataset_size(self, df, name):
        print(f"{name} dataset size: {df.count(), len(df.columns)}")

    def read_data(self, hdfs_path):
        return self.spark.read.parquet(hdfs_path)

    def train_xgboost_model(self, train_data, features_col, label_col, num_workers, params, model_path):
        clf = SparkXGBClassifier(features_col=features_col, label_col=label_col, num_workers=num_workers, **params)
        clf.setParams(**params)
        model = clf.fit(train_data)
        model.write().overwrite().save(model_path)
        return model

    def make_predictions(self, model, input_data):
        return model.transform(input_data)

    def save_predictions_to_csv(self, predictions, output_path):
        predictions.write.mode("overwrite").csv(output_path)

    def get_psi_buckets(self, df, variable):
        """
        Calculate PSI (Population Stability Index) buckets for a given variable.
        PSI measures the distributional change between two datasets.

        Parameters:
        - df: Spark DataFrame containing the variable and weights.
        - variable: Name of the variable for which PSI is calculated.

        Returns:
        - psi_buckets: Dictionary containing PSI buckets.
        """
        psi_buckets = {}
        start = time.time()
        print("This is the iteration of variable:", variable)

        y = [variable, 'weight1']
        trainData_sel = df.select(y)
        trainData_sel = trainData_sel.withColumn('row_nb', F.row_number().over(Window.orderBy(trainData_sel[variable])))
        trainData_sel = trainData_sel.withColumn('cumsum', F.expr('sum(weight1) over (order by row_nb)'))
        train_cum_percent = trainData_sel.withColumn('cum_percent', F.col('cumsum') / F.sum(trainData_sel.weight1).over(Window.partitionBy()))

        bins = [0.1 * i for i in range(11)]
        train_cum_bin = train_cum_percent.withColumn("bin", F.when((F.col('cum_percent') < bins[0]), F.lit(1))
                                                     .otherwise(*[
                                                         F.when((F.col('cum_percent') >= bins[i]) & (F.col('cum_percent') < bins[i + 1]), F.lit(i + 2))
                                                         for i in range(9)
                                                     ])
                                                     .when((F.col('cum_percent') >= bins[9]), F.lit(11))
                                                     )

        train_cum_min_max = train_cum_bin.groupby('bin').agg(
            F.min(F.col(variable)).alias('min_val'),
            F.max(F.col(variable)).alias('max_val'),
            F.sum(F.col('weight1')).alias('Volume')
        )

        train_min_max = train_cum_min_max.select(
            'min_val',
            'max_val',
            'Volume'
        ).orderBy('min_val')

        trainData_N_Obs = train_min_max.agg(F.sum('Volume')).collect()[0][0]
        train_min_max = train_min_max.withColumn('train_per_obs', F.col('Volume') / trainData_N_Obs)

        psi_buckets[variable] = [list(train_min_max.select('min_val').toPandas()['min_val']),
                                 list(train_min_max.select('max_val').toPandas()['max_val'])]

        print("Calculation of PSI Buckets Takes:", time.time() - start)
        return psi_buckets

    def bivariate_analysis(self, df, variable, bin_dict):
        """
        Perform bivariate analysis for a given variable.

        Parameters:
        - df: Spark DataFrame containing the predictions, actual fraud, and weights.
        - variable: Name of the variable for bivariate analysis.
        - bin_dict: Dictionary containing PSI buckets for the variable.

        Returns:
        - bin_value: Dictionary containing bivariate analysis results.
        """
        bin_value = {}
        start = time.time()
        print("This is the iteration of variable:", variable)

        min_range = [float(x) for x in bin_dict[variable][0]]
        max_range = [float(x) for x in bin_dict[variable][1]]

        df_temp = df.select(variable, 'fraud', 'prediction', 'weight1', 'md_tran_amt1')
        for i in range(len(min_range)):
            filtered_df = df_temp.filter((F.col(variable) >= min_range[i]) & (F.col(variable) <= max_range[i]))

            actual_bad = (F.sum(filtered_df['fraud'] * filtered_df['weight1'])).alias('actual_bad')
            actual_good = (F.sum(filtered_df['weight1']) - actual_bad).alias('actual_good')

            predicted_bad = (F.sum(filtered_df['prediction'] * filtered_df['weight1'])).alias('predicted_bad')
            predicted_good = (F.sum(filtered_df['weight1']) - predicted_bad).alias('predicted_good')

            actual_bad_rate = actual_bad / (F.sum(filtered_df['weight1'])).alias('actual_bad_rate')
            predicted_bad_rate = predicted_bad / (F.sum(filtered_df['weight1'])).alias('predicted_bad_rate')

            bin_value[F.concat(F.lit(str(min_range[i]) + "-" + str(max_range[i])))] = [
                actual_bad_rate,
                predicted_bad_rate
            ]

        print("Calculation of Bivariate Analysis Takes:", time.time() - start)
        return bin_value

    def plot_bivariate_analysis(self, keys, x_values, y_values, variable):
        """
        Plot bivariate analysis results.

        Parameters:
        - keys: List of keys (bin names).
        - x_values: List of actual bad rates.
        - y_values: List of predicted bad rates.
        - variable: Name of the variable for plotting.
        """
        plt.figure(figsize=(12, 8))
        plt.plot(keys, x_values, 'o-', color='blue', markersize=8, label='Actual bad Rate')
        plt.plot(keys, y_values, '^', color='orange', linestyle='dotted', markersize=8, label='Predicted bad Rate')

        plt.xlabel(variable)
        plt.ylabel('Bad rate')
        plt.title('Bivariate plot')

        plt.xticks(rotation=45)
        for i, key in enumerate(keys):
            plt.text(keys[i], x_values[i], key, ha='right', va='bottom')

        plt.legend()
        plt.show()

    def run_analysis(self):
        # Read data
        hdfs_path = "/tenants/rqd/fraud/sna/mrg_share/PRISMDLV3/"
        train_data = self.read_data(hdfs_path + "PRISMDLV3_model2306_train.parquet")

        # Display dataset size
        self.display_dataset_size(train_data, "Train")

        # Train XGBoost model
        features_col = "features"
        label_col = "fraud"  # Update with your actual label column
        num_workers = 20
        params = {'colsample_bytree': 0.7, 'gamma': 1, 'learning_rate': 0.08, 'max_depth': 8, 'n_estimators': 2000,
                  'reg_alpha': 0, 'reg_lambda': 10, 'scale_pos_weight': 0.25, 'subsample': 0.8}
        model_path = "/tenants/rqd/fraud/mrgr/r729228/prism_dl_v3/model/final_model/avinash_dev_340var_devprams/"
        trained_model = self.train_xgboost_model(train_data, features_col, label_col, num_workers, params, model_path)

        # Make predictions
        oot_val_data = self.read_data(hdfs_path + "PRISMDLV3_model2306_ootVal.parquet")
        predictions_oot_val = self.make_predictions(trained_model, oot_val_data)

        # Save predictions to CSV
        predictions_output_path = "/data/rqd/home/r729228/mid_63706_prismdlv3/2023_Review/model_files/avinash_dev_340var_devprams_bivariate.csv"
        self.save_predictions_to_csv(predictions_oot_val, predictions_output_path)

        # Bivariate analysis
        var_list = ['visa_score', 'prspend_auth_appr_fstauth_540days', 'merchid_apprfrd_rt_30days',  # Add your variables
                    'mcc_zip5_frd_rt_180days', 'zip_dist', 'cnt_frdaprv', 'amt_avg_frdnonaprv']
        bin_dict = self.get_psi_buckets(train_data, var_list[0])  # Assuming the first variable for PSI buckets
        variable_bivariate = var_list[0]  # Assuming the first variable for bivariate analysis

        # Bivariate analysis for the first variable
        my_dict_tune = self.bivariate_analysis(predictions_oot_val, variable_bivariate, bin_dict)
        my_dict_val = self.bivariate_analysis(predictions_oot_val, variable_bivariate, bin_dict)
        my_dict_intimeval = self.bivariate_analysis(predictions_oot_val, variable_bivariate, bin_dict)

        # Plot bivariate analysis
        keys = list(my_dict_val.keys())
        x_values = [my_dict_val[key][0] for key in keys]
        y_values = [my_dict_val[key][1] for key in keys]
        self.plot_bivariate_analysis(keys, x_values, y_values, variable_bivariate)

# Create an instance of the class and run the analysis
analysis_instance = BivariateAnalysis(app_name="your_application_name")
analysis_instance.run_analysis()
